### PR TITLE
UCP/CORE: Release request ID in send state ff

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Release request ID in send state fast-forwarding.

## Why ?

To avoid possible unreleased IDs for requests which allocated their IDs.

## How ?

Check request's ID if `uct_comp::func` is set to some completion callback (except `ucp_ep_flush_completion()`) in `ucp_request_send_state_ff()`

### CODE REVIEW

File: src/ucp/core/ucp_request.c [hunk 1]
```diff
@@ -415,6 +415,11 @@ void ucp_request_send_state_ff(ucp_request_t *req, ucs_status_t status)
                   ucs_status_string(status));
 
     if (req->send.uct.func == ucp_proto_progress_am_single) {
+        ucs_assert(req->send.state.uct_comp.count == 0);
+
+        if (req->id != UCP_REQUEST_ID_INVALID) {
+            ucp_request_id_release(req);
+        }
```
Comment (yosefe) [inline src/ucp/core/ucp_request.c:422]: maybe do it for all requests? e.g before/after this if

Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:422]: We can’t do it in a common code. No need to check completion counter for requests which don’t set completion callback.

Comment (yosefe) [reply src/ucp/core/ucp_request.c:422]: do you mean we will release id which was not allocated? or later on the completion function will try to release an invalid id?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:422]: No, I meant that some of requests don’t have uct_comp::count set to 0 (i.e. requests which don’t set completion callback - e.g. AM a copy) when we do fast-forward. So l, we can’t touch counter to check whether we need to call completion callback or not.

Comment (yosefe) [reply src/ucp/core/ucp_request.c:422]: but, can we do ` if (req->id != UCP_REQUEST_ID_INVALID) { ucp_request_id_release }` in any case?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:422]: Unfortunately, no.
Since we should do it of completion counter is 0

Comment (yosefe) [reply src/ucp/core/ucp_request.c:422]: why so? it's not like we're going to send or receive anything on this request?

Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:422]: yes, but we should complete the request if it reaches comp_count==0 to avoid use after release

Comment (dmitrygx) [reply src/ucp/core/ucp_request.c:422]: @yosefe sorry if I was unclear.
I meant that UCP reset state for each request in `ucp_request_send_state_reset()`, but only for requests that use UCT completion, UCP resets completion counter to `0`.
https://github.com/openucx/ucx/blob/5b5a98a522ce91171c787a4c158ebb7f45703b96/src/ucp/core/ucp_request.inl#L398

For some operations in `ucp_request_send_state_ff()` we should check if completion counter reaches `0` and then complete them. So, we have to identify what the protocol is used by checking its completion callback and then complete the request if completion counter is 0.

Comment (yosefe) [reply src/ucp/core/ucp_request.c:422]: per offline discussion, make ucp_request_id_release() work even if req->id==UCP_REQUEST_ID_INVALID (as nop)

**Comment by dmitrygx** [comment]
@brminich @hoopoepg could you review pls?

File: src/ucs/datastruct/ptr_map.inl [hunk 2]
```diff
@@ -149,9 +150,10 @@ ucs_ptr_map_get(ucs_ptr_map_t *map, ucs_ptr_map_key_t key, int extract,
  * @param [in]  map     Container.
  * @param [in]  key     Key to object pointer.
  *
- * @return       - UCS_OK on success
- *               - UCS_ERR_NO_ELEM if the key is not found in the internal hash
- *                 table.
+ * @return - UCS_OK on success
+ *         - UCS_ERR_NO_PROGRESS if this key is direct and no additional actions
+ *           are required from an upper-layer protocol
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_map.inl:155]: pls document in terms of UCS and not UCP
for example:
UCS_ERR_NO_PROGRESS if this key is direct and therefore no action was performed
(also for other funcs)

Comment (dmitrygx) [reply src/ucs/datastruct/ptr_map.inl:155]: But we don't do any actions in UCS in case of indirect key. They are performed on an upper-layer (e.g. UCP)

Comment (yosefe) [reply src/ucs/datastruct/ptr_map.inl:155]: that's right.. the return value means no action has been done by ptr_map. and that's it. no need to mention how "upper layer" should /should not handle it.

Comment (dmitrygx) [reply src/ucs/datastruct/ptr_map.inl]: fixed

**Review by yosefe — APPROVED** [review]

**Pull request merged**